lib/racecar/consumer_set.rb

```diff
@@ -0,0 +1,177 @@
+module Racecar
+  class ConsumerSet
+    def initialize(config, logger)
+      @config, @logger = config, logger
+      raise ArgumentError, "Subscriptions must not be empty when subscribing" if @config.subscriptions.empty?
+
+      @consumers = []
+      @consumer_id_iterator = (0...@config.subscriptions.size).cycle
+
+      @last_poll_read_nil_message = false
+    end
+
+    def poll(timeout_ms)
+      maybe_select_next_consumer
+      retried ||= false
+      msg = current.poll(timeout_ms)
+    rescue Rdkafka::RdkafkaError => e
+      @logger.error "Error for topic subscription #{current_subscription}: #{e}"
+
+      case e.code
+      when :max_poll_exceeded, :"err_-147?" # Note: requires unreleased librdkafka version
+        reset_current_consumer
+        raise if retried
+        retried = true
+        retry
+      else
+        raise
+      end
+    ensure
+      @last_poll_read_nil_message = true if msg.nil?
+    end
+
+    # XXX: messages are not guaranteed to be from the same partition
```

Comment:
I think that could break existing consumers.

Reply:
Oh, it certainly did. However, `ConsumerSet` is an entirely new class and we fixed this issue in `Racecar::Runner` in this commit: https://github.com/zendesk/racecar/pull/97/commits/c8902cff80cfe63892730277d4e2bf9372b04516#diff-2b2ca276543198a74d7942069ba0b48bR45

On `master` `Racecar::Runner#consumer` is public, but in this branch it's actually private, so it would break existing custom users already I think. Additionally the methods are named `each_batch` (master) and `batch_poll` (this branch).

I'm aware the behaviour change is not ideal, especially because it is potentially performance degrading (see linked commit message). If someone decides to use the `ConsumerSet#batch_poll` directly, they might be in for a surprise because it's not necessarily visible in testing. Naming the method something like `batch_poll_from_subscribed_partitions` would make it clearer, but it's not an exactly nice name. I'm fine with renaming this because it's not used often, but ideally this would be solved directly within librdkafka.

Reply:
@thijsc would this be something that librdkafka could handle?

Reply:
I asked in `librdkafka`'s channel if there is a recommended approach to handle this in general (https://gitter.im/edenhill/librdkafka?at=5d4d71d21db76d0ffea12263). Let's see what they recommend, maybe I'm just missing some obvious batch poll method that supports this more or less out of the box.

Reply:
from the chat:
> @breunigs what you can do is extract the per-partition queue, see queue_get_partition() and forward it to NULL and poll each partition queue explicitly.

Header for `queue_get_partition` is here: https://github.com/edenhill/librdkafka/blob/675e764b0fb013835c3ead74415bdb62b0aa7903/src/rdkafka.h#L2859-L2872

I took a brief look and this seems to allow a very fine grained setup of how message are consumed. Since there's a rebalance callback, old queues can probably be destroyed through that, but maybe `queue_get_partition` → `poll` → `queue_destroy` in the main loop is easier.
librdkafka also works on all assigned partitions semi-simultaneously, so we'd need something similar to switch queues if we stall on a single partition, or introduce a max time to ensure partitions are worked off more or less evenly. There'd also need to be a call to determine how many potential partitions a topic has, or at least how many we're currently subscribed to.

Apart from handling the details, exposing this function through rdkafka-ruby sounds reasonable, but reimplementing the polling/queue switching handling is probably on racecar side. 

Judging from gut feeling only, this seems effort and not a good trade-off. It took me a couple of iterations until I had all the bugs worked out on another feature which works with the lower-level APIs because it requires a very good understanding of Kafka and librdkafka. However, for us this non-optimal batching also incurs only a small performance penalty, so it was much easier to just start an additional process.